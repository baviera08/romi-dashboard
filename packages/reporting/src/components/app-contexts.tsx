import React from 'react';
import appConfig, { AppConfig } from '../app-config';
// import { defaultSettings, Settings } from '../settings';

// export const SettingsContext = React.createContext(defaultSettings());

export const AppConfigContext = React.createContext<AppConfig>(appConfig);
