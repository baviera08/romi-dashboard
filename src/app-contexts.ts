import { createContext } from "react";
import { IconConfigurationsType } from "./icons-manager";

/* Declares the IconContext which contains the icons used on the app*/
export const IconContext = createContext<IconConfigurationsType>({});
