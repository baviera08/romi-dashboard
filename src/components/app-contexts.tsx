import React from 'react';
import ResourceManager, { ResourceConfigurationsType } from '../resource-manager';
import { defaultSettings, Settings } from '../settings';
import { NotificationBarContext, NotificationBarProps } from './notification-bar';

/* Declares the ResourcesContext which contains the resources used on the app*/
export const ResourcesContext = React.createContext<ResourceManager>({} as ResourceManager);

export const SettingsContext = React.createContext(defaultSettings());

export interface AppContextProviderProps extends React.PropsWithChildren<{}> {
  settings: Settings;
  resources?: Partial<ResourceConfigurationsType>;
  notificationDispatch: React.Dispatch<React.SetStateAction<NotificationBarProps | null>>;
}

export function AppContextProvider(props: AppContextProviderProps): React.ReactElement {
  const { settings, notificationDispatch, resources, children } = props;
  return (
    <SettingsContext.Provider value={settings}>
      <NotificationBarContext.Provider value={notificationDispatch}>
        <ResourcesContext.Provider
          value={
            !!resources?.robots
              ? new ResourceManager(resources as ResourceConfigurationsType)
              : ({} as ResourceManager)
          }
        >
          {children}
        </ResourcesContext.Provider>
      </NotificationBarContext.Provider>
    </SettingsContext.Provider>
  );
}
