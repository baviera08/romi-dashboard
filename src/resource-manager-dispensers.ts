import { RobotResourceManager } from './resource-manager-robots';

interface Location {
  x: number;
  y: number;
  yaw: number;
  level_name: string;
}

export interface ResourceDispenserConfigurationInterface {
  icons: Record<string, string>;
  location: Location;
  name: string;
}

export type ResourceDispenserConfigurationType = Record<
  string,
  ResourceDispenserConfigurationInterface
>;

export class DispenserResourceManager {
  dispensers: ResourceDispenserConfigurationType;
  robots: RobotResourceManager;

  constructor(
    dispenserResources: ResourceDispenserConfigurationType,
    robotsInstance: RobotResourceManager,
  ) {
    this.dispensers = dispenserResources;
    this.robots = robotsInstance;
  }

  getAvailableDispenserPerPlace = (fleetName: string, placeName: string) => {
    return this.robots.getDispensersPerFleet(fleetName, placeName);
  };

  getIconPath = (dispenserName: string): string | null => {
    if (!this.dispenserExists(dispenserName)) {
      return null;
    }
    if (!this.dispensers[dispenserName].hasOwnProperty('icons')) {
      return null;
    }
    const rootIconPath = '/assets/icons';
    const robotIcons = this.dispensers[dispenserName].icons[dispenserName];
    return `${rootIconPath}${robotIcons}`;
  };

  dispenserExists = (dispenserName: string) => {
    if (!this.dispensers.hasOwnProperty(dispenserName)) {
      return false;
    }
    return true;
  };

  get all(): ResourceDispenserConfigurationType {
    return this.dispensers;
  }

  get allValues(): ResourceDispenserConfigurationInterface[] {
    return Object.values(this.dispensers);
  }
}
