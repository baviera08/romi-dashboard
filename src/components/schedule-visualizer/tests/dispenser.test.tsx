import { createMount } from '@material-ui/core/test-utils';
import React from 'react';
import fakeResources from '../../../mock/data/resources';
import ResourceManager, { ResourceConfigurationsType } from '../../../resource-manager';
import Dispenser from '../dispenser';
import fakeDispenserStates from '../../../mock/data/dispenser-states';
import { ResourcesContext } from '../../app-contexts';
import toJson from 'enzyme-to-json';
import { DispenserResourceManager } from '../../../resource-manager-dispensers';

const mount = createMount();

describe('Dispenser component', () => {
  let dispenserHandler: DispenserResourceManager;

  beforeEach(() => {
    const dispenserResources = fakeResources().dispensers;
    if (!dispenserResources || !Object.keys(dispenserResources).length) {
      throw new Error('To execute this test you need to instantiate a dispenser');
    }
    dispenserHandler = new DispenserResourceManager(dispenserResources);
  });

  test('Fires click event when it has a state', () => {
    let clicked = false;

    const dispenserStates = fakeDispenserStates();
    const dispensers = dispenserHandler.allValues;
    const root = mount(
      <svg>
        <Dispenser
          dispenser={dispensers[0]}
          dispenserHandler={dispenserHandler}
          footprint={0.5}
          onClick={() => (clicked = true)}
          dispenserState={dispenserStates[dispensers[0].guid]}
        />
      </svg>,
    );

    root.find(Dispenser).simulate('click');
    expect(clicked).toBe(true);

    root.unmount();
  });

  test(`Don't Fire click event when it has no state`, () => {
    let clicked = false;
    const dispensers = dispenserHandler.allValues;
    const root = mount(
      <svg>
        <Dispenser
          dispenser={dispensers[0]}
          footprint={0.5}
          onClick={() => (clicked = true)}
          dispenserHandler={dispenserHandler}
        />
      </svg>,
    );

    root.find(Dispenser).simulate('click');
    expect(clicked).toBe(false);

    root.unmount();
  });

  test(`Renders correctly`, () => {
    const dispensers = dispenserHandler.allValues;
    const wrapper = mount(
      <svg>
        <Dispenser dispenser={dispensers[0]} footprint={0.5} dispenserHandler={dispenserHandler} />
      </svg>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
