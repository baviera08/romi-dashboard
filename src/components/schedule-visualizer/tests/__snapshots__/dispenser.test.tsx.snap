// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dispenser component Renders correctly 1`] = `
<svg>
  <ForwardRef
    dispenser={
      Object {
        "guid": "coke_ingestor",
        "icons": Object {
          "coke_ingestor": "",
        },
        "location": Object {
          "level_name": "L1",
          "x": 20.95,
          "y": -6.95,
          "yaw": 0,
        },
      }
    }
    dispenserHandler={
      DispenserResourceManager {
        "dispenserExists": [Function],
        "dispensers": Object {
          "coke_dispenser": Object {
            "guid": "coke_dispenser",
            "icons": Object {
              "coke_dispenser": "",
            },
            "location": Object {
              "level_name": "L1",
              "x": 16.85,
              "y": -4.8,
              "yaw": 0,
            },
          },
          "coke_ingestor": Object {
            "guid": "coke_ingestor",
            "icons": Object {
              "coke_ingestor": "",
            },
            "location": Object {
              "level_name": "L1",
              "x": 20.95,
              "y": -6.95,
              "yaw": 0,
            },
          },
        },
        "getIconPath": [Function],
      }
    }
    footprint={0.5}
  >
    <g
      className="makeStyles-container-6"
      onClick={[Function]}
      transform="translate(20.95 6.95)
        rotate(0)"
    >
      <g
        transform="translate(-0.5 -0.5)"
      >
        <image
          height={1}
          href="/assets/move_to_inbox_black_192x192.png"
          onError={[Function]}
          width={1}
        />
      </g>
    </g>
    <SvgText
      className="makeStyles-dispenserText-5"
      id="dispenserName"
      targetWidth={1.1}
      text="coke_ingestor"
      x={20.95}
      y={6.95}
    >
      <text
        className="makeStyles-dispenserText-5"
        id="dispenserName"
        x={20.95}
        y={6.95}
      />
    </SvgText>
  </ForwardRef>
</svg>
`;
